src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Debugging/DefaultRazorProximityExpressionResolver.cs

```diff
@@ -0,0 +1,110 @@
+    {
+        private readonly FileUriProvider _fileUriProvider;
+        private readonly LSPDocumentManager _documentManager;
+        private readonly LSPProjectionProvider _projectionProvider;
+        private readonly CSharpProximityExpressionResolver _csharpProximityExpressionResolver;
+
+        [ImportingConstructor]
+        public DefaultRazorProximityExpressionResolver(
+            FileUriProvider fileUriProvider,
+            LSPDocumentManager documentManager,
+            LSPProjectionProvider projectionProvider,
+            CSharpProximityExpressionResolver csharpProximityExpressionResolver)
+        {
+            if (fileUriProvider is null)
+            {
+                throw new ArgumentNullException(nameof(fileUriProvider));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (projectionProvider is null)
+            {
+                throw new ArgumentNullException(nameof(projectionProvider));
+            }
+
+            if (csharpProximityExpressionResolver is null)
+            {
+                throw new ArgumentNullException(nameof(csharpProximityExpressionResolver));
+            }
+
+            _fileUriProvider = fileUriProvider;
+            _documentManager = documentManager;
+            _projectionProvider = projectionProvider;
+            _csharpProximityExpressionResolver = csharpProximityExpressionResolver;
+        }
+
+        public override async Task<IReadOnlyList<string>> TryResolveProximityExpressionsAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+        {
+            if (textBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            if (!_fileUriProvider.TryGet(textBuffer, out var documentUri))
+            {
+                // Not an addressable Razor document. Do not allow expression resolution here. In practice this shouldn't happen, just being defensive.
+                return null;
+            }
+
+            if (!_documentManager.TryGetDocument(documentUri, out var documentSnapshot))
+            {
+                // No associated Razor document. Do not resolve expressions here. In practice this shouldn't happen, just being defensive.
+                return null;
+            }
+
+            var lspPosition = new Position(lineIndex, characterIndex);
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, lspPosition, cancellationToken).ConfigureAwait(false);
+            if (projectionResult == null)
+            {
+                // Can't map the position, invalid breakpoint location.
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
🔥 

Just once at top of method?

Reply:
Eh you could but checking for cancellation after actual async work is typically more correct. Reason being, if a caller is calling into you with a cancellation token it's up to them to determine if the token is actually cancelled or not.